fdbserver/BlobWorker.actor.cpp

```diff
@@ -258,6 +280,83 @@ static void checkGranuleLock(int64_t epoch, int64_t seqno, int64_t ownerEpoch, i
+                                                                  KeyRange keyRange,
+                                                                  Arena* arena) {
+	state BlobGranuleCipherKeysCtx cipherKeysCtx;
+	state Reference<GranuleTenantData> tenantData = bwData->tenantData.getDataForGranule(keyRange);
+
+	ASSERT(tenantData.isValid());
+
+	std::unordered_map<EncryptCipherDomainId, EncryptCipherDomainName> domains;
+	domains.emplace(tenantData->entry.id, StringRef(*arena, tenantData->name));
+	std::unordered_map<EncryptCipherDomainId, Reference<BlobCipherKey>> domainKeyMap =
+	    wait(getLatestEncryptCipherKeys(bwData->dbInfo, domains));
+
+	auto domainKeyItr = domainKeyMap.find(tenantData->entry.id);
+	ASSERT(domainKeyItr != domainKeyMap.end());
+	cipherKeysCtx.textCipherKey = BlobGranuleCipherKey::fromBlobCipherKey(domainKeyItr->second, *arena);
+
+	TextAndHeaderCipherKeys systemCipherKeys = wait(getLatestSystemEncryptCipherKeys(bwData->dbInfo));
+	cipherKeysCtx.headerCipherKey = BlobGranuleCipherKey::fromBlobCipherKey(systemCipherKeys.cipherHeaderKey, *arena);
+
+	cipherKeysCtx.ivRef = makeString(AES_256_IV_LENGTH, *arena);
+	generateRandomData(mutateString(cipherKeysCtx.ivRef), AES_256_IV_LENGTH);
+
+	TraceEvent("GetLatestGranuleCipherKey")
+	    .detail("IVChksum", XXH3_64bits(cipherKeysCtx.ivRef.begin(), cipherKeysCtx.ivRef.size()));
+
+	return cipherKeysCtx;
+}
+
+ACTOR Future<BlobGranuleCipherKey> lookupCipherKey(Reference<BlobWorkerData> bwData,
+                                                   BlobCipherDetails cipherDetails,
+                                                   Arena* arena) {
+	std::unordered_set<BlobCipherDetails> cipherDetailsSet;
+	cipherDetailsSet.emplace(cipherDetails);
+	state std::unordered_map<BlobCipherDetails, Reference<BlobCipherKey>> cipherKeyMap =
+	    wait(getEncryptCipherKeys(bwData->dbInfo, cipherDetailsSet));
+
+	ASSERT(cipherKeyMap.size() == 1);
+
+	auto cipherKeyMapItr = cipherKeyMap.find(cipherDetails);
+	if (cipherKeyMapItr == cipherKeyMap.end()) {
+		TraceEvent(SevError, "CipherKeyLookup_Failure")
+		    .detail("EncryptDomainId", cipherDetails.encryptDomainId)
+		    .detail("BaseCipherId", cipherDetails.baseCipherId)
+		    .detail("Salt", cipherDetails.salt);
+		throw encrypt_keys_fetch_failed();
+	}
+
+	return BlobGranuleCipherKey::fromBlobCipherKey(cipherKeyMapItr->second, *arena);
+}
+
+ACTOR Future<BlobGranuleCipherKeysCtx> getGranuleCipherKeys(Reference<BlobWorkerData> bwData,
+                                                            BlobGranuleCipherKeysMetaRef cipherKeysMetaRef,
+                                                            Arena* arena) {
+	state BlobGranuleCipherKeysCtx cipherKeysCtx;
+
+	// Fetch 'textCipher' key
+	state BlobCipherDetails textCipherDetails(
+	    cipherKeysMetaRef.textDomainId, cipherKeysMetaRef.textBaseCipherId, cipherKeysMetaRef.textSalt);
+	BlobGranuleCipherKey textCipherKey = wait(lookupCipherKey(bwData, textCipherDetails, arena));
```

Comment:
If this operation is latency sensitive we should consider parallelizing or batching these lookups

Reply:
To add more context, below are relevant details:

Two _encryption key_ types are supported:
1. Revocable keys
2. Non-revocable keys 

External KMS (KeyManagementServer) determines if an encyrption key is `revocable`, if yes, then the FDB processes can cache it (in-memory) for specified (KMS defined) time interval. For instance: `latest encryption keys` are time-bound to support KMS key-rotation; however, point-lookups (rotated EKs) _may_ have not time bound

`BlobCipherKeyCache` is a process local in-memory cache for EKs fetched from EKP (and/or KMS). Lookup actor first lookup local cache to find a _valid_ EK (if present). If not, then it reached out to EKP, which in-turn maintain its own local in-memory cache (KMS BaseCipher cache), however, if desired key is _not found_ or _expired_, then external KMS call gets invoked. 

Further, _latest encryption keys_ gets refreshed by EKP using a _timer task_ to avoid any delays to _inline write path_ (frequency higher than EK TTL), hence, unless KMS is unreachable, _latest EK fetch_ should be serviced at-most by reaching to EKP. For reads, KMS calls may be involved given key _may not be cached_ by FDB cluster OR EKP crash/restarts. 

Overall, I feel the very first EK fetch (for reads) may be expensive, but then subsequent lookups mostly should be a in-process map lookup OR RPC to EKP server.

Reply:
Yes in general this should be very fast, but having things that are very fast most of the time and slow occasionally is how you get bad tail latencies haha. IMO that's why it is still worth optimizing the worst case even if it's not the common case.
If it is likely to become or does become a performance problem, we can always address this in a separate PR though.

Reply:
Yup, lets profile the code running some experiments; it should give us datapoints for better optimization as well as avoid over-engineering if not needed :). 

In some ways, if keys fetches are blocked behind KMS calls, nothing much could be done.

Reply:
> slow occasionally is how you get bad tail latencies
Just to clarify: for _writes_, EKP should have updated EKs unless KMS goes unavailable for a significant time. For reads, KMS needs to be contacted incase the desired EK isn't cached in FDB (local or EKP cache). Also, the TTL for EK is governed by KMS and could be programmed to be substantially longer based on how frequency EK are rotated by the given KMS. 
Point I want to drive is, KMS calls (longer latency) should be only for scenarios such as: _cold reads_ , _crash/restart_ etc; otherwise EK must be fetched once and served from local/EKP cache almost all the time.